virtualenv.py

```diff
@@ -1185,26 +1185,55 @@ def change_prefix(filename, dst_prefix):
     assert False, "Filename {} does not start with any of these prefixes: {}".format(filename, prefixes)
 
 
-def copy_required_modules(dst_prefix, symlink):
-    import warnings
+if sys.version_info < (3, 4):
+    import imp
+
+    def find_module_filename(modname):
+        try:
+            f, filename, _ = imp.find_module(modname)
+        except ImportError:
+            return None
+        else:
+            if f is not None:
+                f.close()
+            return filename
+
 
-    with warnings.catch_warnings():
-        # Ignore deprecation of the imp module
-        # TODO: do not use deprecated imp module
-        warnings.simplefilter("ignore")
-        import imp
+else:
+    import importlib.util
+
+    if sys.version_info < (3, 5):
 
+        def find_spec(modname):
+            loader = importlib.find_loader(modname)
+            if loader is None:
+                return None
+            else:
+                return importlib.util.spec_from_loader(modname, loader)
+
+    else:
+        find_spec = importlib.util.find_spec
+
+    def find_module_filename(modname):
+        spec = importlib.util.find_spec(modname)
+        # on pypy3, some builtin modules havae a bogus build-time file path
+        if spec is None or not os.path.exists(spec.origin):
+            return None
+        elif os.path.basename(spec.origin) == "__init__.py":
```

Comment:
I would prefer here to set filename as spec.origin and have an if to fix up when it's just ``__init__.py`` 🤔

Reply:
can it be non py ending? (pyc, .class (jython), etc?)

Reply:
unclear how jython will implement this since jython is stuck in python2.x -- should always be `.py` as far as I know as well (it's a SourceFileLoader)

Reply:
Does the source file loader make this guarantee in the documentation, then?

Reply:
specified here: https://docs.python.org/3/library/importlib.html#importlib.machinery.ModuleSpec as equivalent to `__file__`

PEP 3147 says that `__file__` points to the _source file_: https://www.python.org/dev/peps/pep-3147/#file (and `__cached__` will refer to the `.pyc` file)